control-plane/connect-inject/peering_dialer_controller_test.go

```diff
@@ -0,0 +1,780 @@
+					SecretRef: &v1alpha1.SecretRefStatus{
+						Secret: v1alpha1.Secret{
+							Name:    "foo",
+							Key:     "token",
+							Backend: "kubernetes",
+						},
+						ResourceVersion: "version1",
+					},
+				},
+			},
+			secret: &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					ResourceVersion: "version1",
+				},
+			},
+			isDifferent: false,
+		},
+	}
+
+	for name, tt := range cases {
+		t.Run(name, func(t *testing.T) {
+			controller := PeeringDialerController{}
+			isDifferent := controller.specStatusSecretsDifferent(tt.dialer, tt.secret)
+			require.Equal(t, tt.isDifferent, isDifferent)
+		})
+	}
+}
+
+// TestReconcileDeletePeeringDialer reconciles a PeeringDialer resource that is no longer in Kubernetes, but still
+// exists in Consul.
+func TestReconcileDeletePeeringDialer(t *testing.T) {
+	// Add the default namespace.
+	ns := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "default"}}
+
+	dialer := &v1alpha1.PeeringDialer{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:              "dialer-deleted",
+			Namespace:         "default",
+			DeletionTimestamp: &metav1.Time{Time: time.Now()},
+			Finalizers:        []string{FinalizerName},
+		},
+		Spec: v1alpha1.PeeringDialerSpec{
+			Peer: &v1alpha1.Peer{
+				Secret: nil,
+			},
+		},
+	}
+
+	// Create fake k8s client.
+	k8sObjects := []runtime.Object{ns, dialer}
+
+	// Add peering types to the scheme.
+	s := scheme.Scheme
+	s.AddKnownTypes(v1alpha1.GroupVersion, &v1alpha1.PeeringDialer{}, &v1alpha1.PeeringDialerList{})
+	fakeClient := fake.NewClientBuilder().WithScheme(s).WithRuntimeObjects(k8sObjects...).Build()
+
+	// Create test consul server.
+	consul, err := testutil.NewTestServerConfigT(t, func(c *testutil.TestServerConfig) {
+		c.NodeName = "test-node"
+	})
+	require.NoError(t, err)
+	defer consul.Stop()
+	consul.WaitForServiceIntentions(t)
+
+	cfg := &api.Config{
+		Address: consul.HTTPAddr,
+	}
+	consulClient, err := api.NewClient(cfg)
+	require.NoError(t, err)
+
+	// Add the initial peerings into Consul by calling the Generate token endpoint.
+	_, _, err = consulClient.Peerings().GenerateToken(context.Background(), api.PeeringGenerateTokenRequest{PeerName: "dialer-deleted"}, nil)
+	require.NoError(t, err)
+
+	// Create the peering dialer controller.
+	pdc := &PeeringDialerController{
+		Client:       fakeClient,
+		Log:          logrtest.TestLogger{T: t},
+		ConsulClient: consulClient,
+		Scheme:       s,
+	}
+	namespacedName := types.NamespacedName{
+		Name:      "dialer-deleted",
+		Namespace: "default",
+	}
+
+	// Reconcile a resource that is not in K8s, but is still in Consul.
+	resp, err := pdc.Reconcile(context.Background(), ctrl.Request{
+		NamespacedName: namespacedName,
+	})
+	require.NoError(t, err)
+	require.False(t, resp.Requeue)
+
+	// After reconciliation, Consul should have a peering with its DeletedAt set.
+	peering, _, err := consulClient.Peerings().Read(context.Background(), "dialer-deleted", nil)
+	require.NotNil(t, peering.DeletedAt)
```

Comment:
Seeing the tests fail with a nil pointer exception here. I think its because the deletion logic was implemented in Consul, and the peering is sometimes being deleted before we can read it here. So I don't think we can make this assertion anymore. Maybe we can do a retry to check that the peering is nil? Since it's a deferred delete?

Reply:
I think you need to build a newer Consul image. I ran into the same issue but turns out, I had an older Consul binary.

Reply:
Hmm, I tried the retry to check that the peering is nil with a timer that waits for 5 seconds, and that also fails.

Reply:
THE DEFERRED DELETION!!! ok ok. that makes sense. thank you so much Nitya! that was an excellent catch.